File: src/juicy/root/js/bridge.js
$(function () {

    bridge.callbacks = {};
    
    bridge.signal.connect(function(raw) {
        var message = JSON.parse(raw);
        if (bridge.callbacks.hasOwnProperty(message.name)) {
            bridge.callbacks[message.name].forEach(function (callback) {
                callback(message);
            });
        }
    });

    bridge.on = function (name, callback) {
        if (!bridge.callbacks.hasOwnProperty(name)) {
            bridge.callbacks[name] = [];
            bridge.listen(name);
        }
        bridge.callbacks[name].push(callback);
    };

    bridge.trigger = function (message) {
        var raw = JSON.stringify(message);
        bridge.send(raw);
    };


});

File: src/juicy/core/bridge.py
import json

from PySide import QtGui
from PySide import QtCore

import juicy


class Bridge(QtCore.QObject):

    signal = QtCore.Signal(str)

    def __init__(self):
        super(Bridge, self).__init__()

    @QtCore.Slot(str)
    def listen(self, name):
        juicy.mq.listen(name, Bridge.on_message)

    @staticmethod
    def on_message(message):
        Bridge.signal.emit(json.dumps(message))

    @QtCore.Slot(str)
    def send(self, raw):
        message = json.loads(raw)
        juicy.mq.send(message)
